### ISSUE TITLE
[A2C强化学习组合优化] | A2C Portfolio Optimization

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4531]
**标签**: quantitative-strategy
### 摘要
使用Advantage Actor-Critic(A2C)强化学习算法优化多资产投资组合配置。

### 算法描述
**核心原理**：
A2C是一种基于策略梯度的强化学习算法，通过同时学习值函数（Critic）和策略函数（Actor）来最大化长期累积收益。在组合优化场景中，Agent根据市场状态动态调整各资产权重。

**数学框架**：
1. **MDP建模**：
   - State: $s_t = (r_{t-1}, r_{t-2}, ..., r_{t-n}, w_{t-1})$
   - Action: $a_t = \pi(s_t)$，资产权重向量
   - Reward: $r_t = \sum_i w_{i,t-1} \cdot r_{i,t}$

2. **Critic (Value Function)**：
   $$V(s_t; \theta_v) = \mathbb{E}[R_t | s_t]$$
   $$L_{value} = \mathbb{E}[(R_t - V(s_t; \theta_v))^2]$$

3. **Actor (Policy)**：
   $$\pi(a_t|s_t; \theta_\pi) = \text{Softmax}(f(s_t; \theta_\pi))$$
   $$L_{policy} = -\mathbb{E}[\log \pi(a_t|s_t; \theta_\pi) \cdot A(s_t, a_t)]$$

4. **Advantage函数**：
   $$A(s_t, a_t) = R_t - V(s_t; \theta_v) + \gamma V(s_{t+1}; \theta_v)$$

5. **熵正则化**：
   $$L_{total} = L_{policy} + \alpha \cdot H(\pi(\cdot|s_t))$$

**策略流程**：
```
1. 初始化Actor-Critic网络
2. 收集市场状态序列
3. 计算 Advantage 估计
4. 更新 Actor 网络 (策略梯度)
5. 更新 Critic 网络 (值函数拟合)
6. 周期性执行步骤2-5
7. 根据学习到的策略生成交易信号
```

**关键参数**：
- `gamma`: 折扣因子（默认0.99）
- `learning_rate`: 学习率（默认3e-4）
- `entropy_coef`: 熵正则化系数（默认0.01）
- `n_steps`: 每次更新步数（默认2048）

### Python代码示例
```python
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from collections import deque


class ActorCriticNetwork(nn.Module):
    def __init__(self, state_dim, action_dim, hidden_dim=256):
        super().__init__()
        self.actor = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, action_dim),
            nn.Softmax(dim=-1)
        )
        self.critic = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    
    def forward(self, x):
        return self.actor(x), self.critic(x)


class A2CPortfolioOptimizer:
    def __init__(
        self,
        state_dim=20,
        n_assets=5,
        gamma=0.99,
        learning_rate=3e-4,
        entropy_coef=0.01,
        n_steps=2048
    ):
        self.state_dim = state_dim
        self.n_assets = n_assets
        self.gamma = gamma
        self.entropy_coef = entropy_coef
        self.n_steps = n_steps
        
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.policy = ActorCriticNetwork(state_dim, n_assets).to(self.device)
        self.optimizer = optim.Adam(self.policy.parameters(), lr=learning_rate)
        
        self.memory = deque(maxlen=10000)
        self.step_count = 0
        
    def get_state(self, returns_history, weights):
        features = []
        for asset_returns in returns_history:
            features.extend([
                np.mean(asset_returns[-20:]),
                np.std(asset_returns[-20:]),
                np.min(asset_returns[-20:]),
                np.max(asset_returns[-20:])
            ])
        features.extend(weights)
        return np.array(features[:self.state_dim])
    
    def choose_action(self, state):
        state_tensor = torch.FloatTensor(state).unsqueeze(0).to(self.device)
        probs, _ = self.policy(state_tensor)
        action = torch.multinomial(probs, 1).item()
        weights = probs.squeeze().cpu().numpy()
        return action, weights
    
    def store_transition(self, state, action, reward, next_state, done):
        self.memory.append({
            'state': state,
            'action': action,
            'reward': reward,
            'next_state': next_state,
            'done': done
        })
        self.step_count += 1
    
    def compute_returns(self, rewards, dones):
        returns = []
        discounted_reward = 0
        for reward, done in zip(reversed(rewards), reversed(dones)):
            discounted_reward = reward + self.gamma * discounted_reward
            returns.insert(0, discounted_reward)
        return np.array(returns)
    
    def update(self):
        if len(self.memory) < self.n_steps:
            return
        
        batch = [self.memory[i] for i in 
                 np.random.choice(len(self.memory), self.n_steps, replace=False)]
        
        states = torch.FloatTensor(np.array([t['state'] for t in batch])).to(self.device)
        actions = torch.LongTensor([t['action'] for t in batch]).to(self.device)
        rewards = torch.FloatTensor([t['reward'] for t in batch]).to(self.device)
        next_states = torch.FloatTensor(np.array([t['next_state'] for t in batch])).to(self.device)
        dones = torch.FloatTensor([t['done'] for t in batch]).to(self.device)
        
        probs, values = self.policy(states)
        _, next_values = self.policy(next_states)
        
        returns = self.compute_returns(rewards.numpy(), dones.numpy())
        returns_tensor = torch.FloatTensor(returns).to(self.device)
        
        advantages = returns_tensor - values.squeeze()
        
        dist = torch.distributions.Categorical(probs)
        log_probs = dist.log_prob(actions)
        entropy = dist.entropy().mean()
        
        policy_loss = -(log_probs * advantages.detach()).mean()
        value_loss = advantages.pow(2).mean()
        
        loss = policy_loss + 0.5 * value_loss - self.entropy_coef * entropy
        
        self.optimizer.zero_grad()
        loss.backward()
        torch.nn.utils.clip_grad_norm_(self.policy.parameters(), 0.5)
        self.optimizer.step()
        
    def calculate_portfolio_return(self, weights, asset_returns):
        return np.dot(weights, asset_returns)
    
    def rebalance(self, current_weights, target_weights, transaction_cost=0.001):
        turnover = np.sum(np.abs(target_weights - current_weights))
        cost = turnover * transaction_cost
        return target_weights, cost


if __name__ == "__main__":
    env = A2CPortfolioOptimizer(state_dim=20, n_assets=5)
    
    np.random.seed(42)
    torch.manual_seed(42)
    
    n_episodes = 100
    portfolio_values = []
    
    for episode in range(n_episodes):
        returns_history = [
            np.random.randn(100) * 0.02 for _ in range(5)
        ]
        current_weights = np.array([0.2, 0.2, 0.2, 0.2, 0.2])
        portfolio_value = 1.0
        
        episode_reward = 0
        for step in range(100):
            state = env.get_state(returns_history, current_weights)
            action, new_weights = env.choose_action(state)
            
            asset_returns = np.array([
                np.random.randn() * 0.02 for _ in range(5)
            ])
            
            portfolio_return = env.calculate_portfolio_return(new_weights, asset_returns)
            portfolio_value *= (1 + portfolio_return)
            
            next_state = env.get_state(
                [np.append(r, asset_returns[i]) for i, r in enumerate(returns_history)],
                new_weights
            )
            
            reward = portfolio_return
            done = step == 99
            
            env.store_transition(state, action, reward, next_state, done)
            episode_reward += reward
            
            current_weights = new_weights
            
            for i in range(5):
                returns_history[i] = np.append(returns_history[i], asset_returns[i])
        
        if env.step_count >= env.n_steps:
            env.update()
        
        portfolio_values.append(portfolio_value)
        if episode % 10 == 0:
            print(f"Episode {episode}, Portfolio Value: {portfolio_value:.4f}")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
